org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/PasteEventHandler.java

```diff
@@ -334,6 +343,77 @@ private static String getEol(String text) {
 		return text.contains("\r\n") ? "\r\n" : "\n";
 	}
 
+	public static String handleFilePasteEvent(String path, String content, IProgressMonitor monitor) throws JavaModelException {
+		String desiredPath = "";
+		final ASTParser parser = ASTParser.newParser(AST.getJLSLatest());
+		parser.setResolveBindings(false);
+		parser.setKind(ASTParser.K_COMPILATION_UNIT);
+		parser.setBindingsRecovery(false);
+		parser.setIgnoreMethodBodies(true);
+		parser.setEnvironment(new String[0], new String[0], null, true);
+		Map<String, String> javaOptions = JavaCore.getOptions();
+		parser.setCompilerOptions(javaOptions);
+		parser.setSource(content.toCharArray());
+		CompilationUnit root = (CompilationUnit) parser.createAST(monitor);
+		// Check for package, import, and/or type declaration to determine if java file
+		if (root != null && (root.getPackage() != null || !root.imports().isEmpty() || !root.types().isEmpty())) {
+			String pathRoot = path;
+			if (root.getPackage() != null) {
+				String packageName = root.getPackage().getName().toString();
+				IJavaProject[] projects = ProjectUtils.getJavaProjects();
+				boolean pathFound = false;
+				String partialPathAddition = "";
+				for (int i = 0; i < projects.length; i++) {
+					if (!ProjectsManager.DEFAULT_PROJECT_NAME.equals(projects[i].getElementName()) && !pathFound) {
+						IPackageFragment[] packages = projects[i].getPackageFragments();
+						for (int j = 0; j < packages.length; j++) {
+							if (packages[j].getKind() == IPackageFragmentRoot.K_SOURCE && !packages[j].isDefaultPackage()) {
+								pathRoot = packages[j].getParent().getResource().getLocation().toOSString();
+								String name = packages[j].getElementName();
+								if (name.equals(packageName)) {
+									desiredPath = packages[j].getResource().getLocation().toOSString();
+									pathFound = true;
+									break;
+								} else if (packageName.startsWith(name) && packageName.substring(0, name.length() + 1).endsWith(".")) {
```

Comment:
I guess this works because `name` is always a proper package, so as long as `packageName` is "completed" by `name` to the `.` it'll be a valid subpackage. Ok fine. There's probably code somwhere to split both terms by `.` and compare each component, but this is nice enough :stuck_out_tongue: